### PR TITLE
[Op] cast, wrap_param

### PR DESCRIPTION
**Pull request by spectrometerHBH**

### CODE REVIEW

File: python/tvm/relax/op/transform.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        The target data type
+    Returns
+    -------
+    result : relax.Expr
+        The casted result.
+    """
+    assert isinstance(data, relax.Constant)
+    if data.data.dtype == dtype:
+        return data
+    if isinstance(dtype, str):
```
Comment (tqchen) [inline python/tvm/relax/op/transform.py]: can directly pass str into FFI, FFI auto converts to DataType

**Comment by MasterJH5574** [comment]
#79 has more complete test coverage for Cast, which you might want to copy-paste to this PR. Also please bring full tests for WrapParam, thanks!

File: python/tvm/script/ir_builder/relax/ir.py [hunk 2]
```diff
@@ -58,6 +58,8 @@
     tanh,
     unique,
     variance,
+    cast,
+    wrap_param,
```
Comment (MasterJH5574) [inline python/tvm/script/ir_builder/relax/ir.py]: Consider reorder them in lexicographical order.

File: python/tvm/script/ir_builder/relax/ir.py [hunk 3]
```diff
@@ -458,4 +460,6 @@ def RewriteSymbolicShape(
     "tensor",
     "unique",
     "variance",
+    "cast",
+    "wrap_param",
```
Comment (MasterJH5574) [inline python/tvm/script/ir_builder/relax/ir.py]: Ditto

File: src/relax/op/tensor/transform.cc [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+TVM_REGISTER_GLOBAL("relax.op.cast").set_body_typed(MakeCast);
+
+StructInfo InferStructInfoCast(const Call& call, const BlockBuilder& ctx) {
+  TensorStructInfo sinfo = GetUnaryInputTensorStructInfo(call, ctx);
+  const auto* attrs = call->attrs.as<CastAttrs>();
+  ObjectPtr<TensorStructInfoNode> new_sinfo = make_object<TensorStructInfoNode>(*sinfo.get());
+  new_sinfo->dtype = attrs->dtype;
+  return TensorStructInfo(new_sinfo);
+}
```
Comment (MasterJH5574) [inline src/relax/op/tensor/transform.cc:49]: We can integrate the two inference function into one, using template.

File: python/tvm/relax/op/transform.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+from typing import Union
+
+import tvm
+from tvm import relax
+
+from . import _ffi_api
+from ..expr import Expr
+
+
+def cast(data: Expr, dtype: Union[str, tvm.DataType]) -> Expr:
```
Comment (MasterJH5574) [inline python/tvm/relax/op/transform.py:27]:
```suggestion
def cast(data: Expr, dtype: str) -> Expr:
```

File: python/tvm/relax/op/transform.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    """Cast input tensor to data type.
+    Parameters
+    ----------
+    data : relax.Expr
+        The input data to the operator.
+    dtype: Union[str, tvm.DataType]
```
Comment (MasterJH5574) [inline python/tvm/relax/op/transform.py:33]:
```suggestion
    dtype: str
```

File: python/tvm/relax/op/transform.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+        The target data type
+    Returns
+    -------
+    result : relax.Expr
+        The casted result.
+    """
+    if isinstance(dtype, str):
+        dtype = tvm.DataType(dtype)
```
Comment (MasterJH5574) [inline python/tvm/relax/op/transform.py:41]:
```suggestion
```

File: python/tvm/relax/op/transform.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    return _ffi_api.cast(data, dtype)  # type: ignore
+
+
+def wrap_param(data: Expr, dtype: Union[str, tvm.DataType] = "float32") -> Expr:
```
Comment (MasterJH5574) [inline python/tvm/relax/op/transform.py:45]:
```suggestion
def wrap_param(data: Expr, dtype: str = "float32") -> Expr:
```

File: python/tvm/relax/op/transform.py [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+    """Cast input tensor which is model param to data type if the dtype of the input data is not
+    the same as the given dtype.
+    Parameters
+    ----------
+    data : relax.Expr
+        The input data to the operator.
+    dtype: Union[str, tvm.DataType]
```
Comment (MasterJH5574) [inline python/tvm/relax/op/transform.py:52]:
```suggestion
    dtype: str
```

**Review by MasterJH5574 — COMMENTED** [review]
str is good. FFI can convert string to corresponding DataType

File: include/tvm/relax/op_attr_types.h [hunk 10]
```diff
@@ -181,6 +181,24 @@ struct UniqueAttrs : public tvm::AttrsNode<UniqueAttrs> {
...
+/*! \brief Attributes used in cast operator */
+struct CastAttrs : public tvm::AttrsNode<CastAttrs> {
+  DataType dtype;
+
+  TVM_DECLARE_ATTRS(CastAttrs, "relax.attrs.CastAttrs") {
+    TVM_ATTR_FIELD(dtype).describe("Target data type");
+  }
+};  // struct CastAttrs.
+
+/*! \brief Attributes used in cast operator */
```
Comment (MasterJH5574) [inline include/tvm/relax/op_attr_types.h]:
```suggestion
/*! \brief Attributes used in wrap_param operator */
```

File: python/tvm/relax/op/transform.py (hunk 5 above)
Comment (spectrometerHBH) [reply python/tvm/relax/op/transform.py:27]: That doesn't mean you should not use tvm.DataType

Comment (spectrometerHBH) [reply python/tvm/relax/op/transform.py:27]:
```
a = ...
b = cast(c, a.dtype)
```

**Review by MasterJH5574 — APPROVED** [review]

**Pull request merged**